File: app/src/main/java/uk/lobsterdoodle/namepicker/storage/SharedPrefsKeyValueStore.java
package uk.lobsterdoodle.namepicker.storage;

import android.content.SharedPreferences;

public class SharedPrefsKeyValueStore implements KeyValueStore {
    private final SharedPreferences sharedPreferences;

    public SharedPrefsKeyValueStore(SharedPreferences sharedPreferences) {
        this.sharedPreferences = sharedPreferences;
    }

    @Override
    public boolean getBool(String key, boolean defaultValue) {
        return sharedPreferences.getBoolean(key, defaultValue);
    }

    @Override
    public long getLong(String key, long defaultValue) {
        return sharedPreferences.getLong(key, defaultValue);
    }

    @Override
    public String getString(String key, String defaultValue) {
        return sharedPreferences.getString(key, defaultValue);
    }

    @Override
    public boolean contains(String key) {
        return sharedPreferences.contains(key);
    }

    @Override
    public Edit edit() {
        return new SharedPrefsEdit(sharedPreferences.edit());
    }

    private class SharedPrefsEdit implements Edit {
        private final SharedPreferences.Editor edit;

        SharedPrefsEdit(SharedPreferences.Editor edit) {
            this.edit = edit;
        }

        @Override
        public Edit put(String key, boolean value) {
            edit.putBoolean(key, value);
            return this;
        }

        @Override
        public Edit put(String key, long value) {
            edit.putLong(key, value);
            return this;
        }

        @Override
        public Edit put(String key, String value) {
            edit.putString(key, value);
            return this;
        }

        @Override
        public Edit remove(String key) {
            edit.remove(key);
            return this;
        }

        @Override
        public Edit clear() {
            edit.clear();
            return this;
        }

        @Override
        public void commit() {
            edit.commit();
        }
    }
}

File: app/src/main/java/uk/lobsterdoodle/namepicker/storage/SharedPrefsKeyValueStore.kt
package uk.lobsterdoodle.namepicker.storage

import android.annotation.SuppressLint
import android.content.SharedPreferences

class SharedPrefsKeyValueStore(private val sharedPreferences: SharedPreferences) : KeyValueStore {

    override fun getBool(key: String, defaultValue: Boolean): Boolean
            = sharedPreferences.getBoolean(key, defaultValue)

    override fun getLong(key: String, defaultValue: Long): Long
            = sharedPreferences.getLong(key, defaultValue)

    override fun getString(key: String, defaultValue: String): String
            = sharedPreferences.getString(key, defaultValue)

    override fun contains(key: String): Boolean
            = sharedPreferences.contains(key)

    @SuppressLint("CommitPrefEdits")
    override fun edit(): KeyValueStore.Edit
            = SharedPrefsEdit(sharedPreferences.edit())

    private inner class SharedPrefsEdit internal constructor(private val edit: SharedPreferences.Editor) : KeyValueStore.Edit {

        override fun put(key: String, value: Boolean): KeyValueStore.Edit {
            edit.putBoolean(key, value)
            return this
        }

        override fun put(key: String, value: Long): KeyValueStore.Edit {
            edit.putLong(key, value)
            return this
        }

        override fun put(key: String, value: String): KeyValueStore.Edit {
            edit.putString(key, value)
            return this
        }

        override fun remove(key: String): KeyValueStore.Edit {
            edit.remove(key)
            return this
        }

        override fun clear(): KeyValueStore.Edit {
            edit.clear()
            return this
        }

        override fun commit() {
            edit.commit()
        }
    }
}
